### DISCUSSION TITLE
2 KY-040 rotary encoder seems to have problem

### DISCUSSION THREAD

**Discussion by heimbastler** [arendst/Tasmota#15430] [Ideas]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
Hi,

I installed 2pcs KY-040 rotary encoder on one nodeMCU. One encoder should control PWM1 the 2nd encoder PWM2. If I install only one encoder everything works well. As I start adding the 2nd things went wrong. I tried many different GPIOs but connect as mentioned in the recomendet list from here: 
[usable pins](https://tasmota.github.io/docs/Peripherals/)

Also I tried to use SetOption68 1 (Splits off the light into individually controlled Channels). This does not work for any channel than. If I switch it off (0) back again only one channel works.

I tried different controller 2 pcs of NodeMCU, GPIO ports, GPIO settings, different encoder, different module types (Generic 0 and 18). Most of the case if I use the 2nd channel the whole module stuck and I have to reflash and setup everything from scrab.

Thanks,
Alex

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [ ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [ ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): NodeMCU
- [ ] Tasmota binary firmware version number used: tasmota.bin 11.1.0 Pre-compiled
  - [ ] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: Tasmotizer
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
{"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
{"Module":{"18":"Generic"}}
{"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"3297":"Rotary B2"},"GPIO5":{"416":"PWM1"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"3264":"Rotary A1"},"GPIO13":{"3296":"Rotary B1"},"GPIO14":{"3265":"Rotary A2"},"GPIO15":{"417":"PWM2"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
15:59:32.277 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS = {"Status":{"Module":18,"DeviceName":"BettRechts","FriendlyName":["BettRechts",""],"Topic":"BettRechts","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
15:59:32.308 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:02:49","StartupUTC":"2022-04-22T14:56:43","Sleep":50,"CfgHolder":4617,"BootCount":92,"BCResetTime":"2022-04-21T14:30:24","SaveCount":395,"SaveAddress":"F9000"}}
15:59:32.346 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS2 = {"StatusFWR":{"Version":"11.1.0(tasmota)","BuildDateTime":"2022-04-13T06:40:42","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"382/699"}}
15:59:32.369 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["nuni",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A190000000000","00040080","00006008","00000000"]}}
15:59:32.404 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS4 = {"StatusMEM":{"ProgramSize":626,"Free":376,"Heap":23,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","04000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,56,62","Sensors":"1,2,3,4,5,6"}}
15:59:32.444 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS5 = {"StatusNET":{"Hostname":"BettRechts-0898","IPAddress":"192.168.178.202","Gateway":"192.168.178.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.178.1","DNSServer2":"0.0.0.0","Mac":"9C:9C:1F:46:23:82","Webserver":2,"HTTP_API":0,"WifiConfig":4,"WifiPower":17.0}}
15:59:32.469 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.178.37","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_462382","MqttUser":"DVES_USER","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
15:59:32.495 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS7 = {"StatusTIM":{"UTC":"2022-04-22T14:59:32","Local":"2022-04-22T15:59:32","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"05:45","Sunset":"19:51"}}
15:59:32.517 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS10 = {"StatusSNS":{"Time":"2022-04-22T15:59:32"}}
15:59:32.529 MQT: SmartHome/Schlafzimmer/BettRechts/stat/STATUS11 = {"StatusSTS":{"Time":"2022-04-22T15:59:32","Uptime":"0T00:02:49","UptimeSec":169,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","Channel1":0,"POWER2":"OFF","Channel2":0,"Color":"0000","Wifi":{"AP":1,"SSId":"nuni","BSSId":"E0:28:6D:70:B2:42","Channel":13,"Mode":"11n","RSSI":58,"Signal":-71,"LinkCount":1,"Downtime":"0T00:00:04"}}}
```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
16:00:28.622 MQT: SmartHome/Schlafzimmer/BettRechts/stat/RESULT = {"POWER1":"OFF","Channel1":0,"POWER2":"OFF","Channel2":0,"Color":"0000"}
16:00:30.869 CFG: Saved to flash at F7, Count 397, Bytes 4096
16:00:31.075 WIF: Checking connection...
16:00:34.883 MQT: SmartHome/Schlafzimmer/BettRechts/stat/RESULT = {"POWER1":"OFF","Channel1":0,"POWER2":"OFF","Channel2":0,"Color":"0000"}
16:00:36.875 CFG: Saved to flash at F6, Count 398, Bytes 4096
16:00:39.789 MQT: SmartHome/Schlafzimmer/BettRechts/stat/RESULT = {"POWER1":"OFF","Channel1":0,"POWER2":"OFF","Channel2":0,"Color":"0000"}
16:00:43.395 MQT: SmartHome/Schlafzimmer/BettRechts/stat/RESULT = {"POWER1":"OFF","Channel1":0,"POWER2":"OFF","Channel2":0,"Color":"0000"}
16:00:43.746 MQT: SmartHome/Schlafzimmer/BettRechts/stat/RESULT = {"POWER1":"OFF","Channel1":0,"POWER2":"OFF","Channel2":0,"Color":"0000"}
16:00:45.347 WIF: Sending Gratuitous ARP
16:00:51.114 WIF: Checking connection...
16:01:01.379 DGR: Sending Board message to network: seq=2, flags=64
16:01:08.195 DGR: Received Board message from 192.168.178.204: seq=14, flags=64
16:01:11.111 WIF: Checking connection...
```

### TO REPRODUCE
_Steps to reproduce the behavior:_
rotate the encoder 1 one step CW
rotate the encoder 1 one step CCW
rotate the encoder 2 one step CW
rotate the encoder 2 one step CCW

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
should set a respective dim parameter to each channel and move the slider in the GUI when browser is refreshed

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
![Bildschirmfoto 2022-04-22 um 17 07 00](https://user-images.githubusercontent.com/103124915/164742478-3200aae3-99b6-4982-8142-05bcdafce24d.png)

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by joeri3000** [comment]
I'm facing kind of the same issue, but not using PWM channels but 2 dimmer channels (MS-105B Tuya device with 2 rotary's attached to free GPIO pins).
When I activate only 1 dimmer channel, the output is like this (in tele/tasmota/state topic):
`"POWER":"OFF","Dimmer":15,"Fade":"OFF","Speed":1`

but when I activate both dimmer channels, the output would be like this:
`"POWER1":"OFF","Channel1":15,"POWER2":"OFF","Channel2":15`

I guess the issue here is that the rotary value seems hardcoded to the "Dimmer" channel which, in a dual channel setup, is replaced to "Channel1" and "Channel2". Therefore turning the rotary doesn't change (inc/dec) the respective dimmer value.

As a workaround I've set up this in rules now:
`ON rotary1#pos1 DO channel1 %value% endon ON rotary2#pos1 DO channel2 %value% endon`
Unfortunately this doesn't keep the rotary value and the dimmer channel synchronized. e.g.: when sending Channel1 to 100 through MQTT followed by turning the rotary one notch down, the value will go to 14 instead of 99 (as per above example).

I suppose a fix could be to check if there is a "Dimmer" channel. If no, map the rotary variables to "Channel1" and "Channel2" respectively (or "PWM1" and "PWM2" in the topic starter's use case).

Hopefully such fix could be implemented in the next release.
